File: backend/app/orchestration/answer_guard.py
from __future__ import annotations

from typing import cast

from app.domain.gate import meaningful_features
from app.domain.models import (
    REQUIRED_FIELDS,
    BriefFieldName,
    ExtractionResult,
    FieldUpdate,
    Session,
)


def _non_empty_text(value: str | list[str] | None) -> bool:
    if isinstance(value, str):
        return bool(value.strip())
    if isinstance(value, list):
        return any(str(item).strip() for item in value)
    return False


def _is_acceptable_price(update: FieldUpdate) -> bool:
    """Accept structured extractor output; do not re-parse natural-language meaning."""
    if update.status == "confirmed":
        return _non_empty_text(update.value)
    return bool((update.raw_text or "").strip())


def _is_acceptable_answer(field: BriefFieldName, update: FieldUpdate) -> bool:
    """True when the update is a real answer for the field we just asked about."""
    if field == "price":
        return _is_acceptable_price(update)
    if update.status != "confirmed":
        return False
    if update.value is None:
        return False
    if field == "key_features":
        return bool(meaningful_features(update.value))
    if isinstance(update.value, str):
        return bool(update.value.strip())
    if isinstance(update.value, list):
        return bool(meaningful_features(update.value))
    return True


def filter_invalid_asked_field_updates(
    session: Session, extraction: ExtractionResult
) -> tuple[ExtractionResult, BriefFieldName | None]:
    """Drop unusable answers to the field we asked for so they never hit the brief.

    Returns the filtered extraction and the rejected field name (if any).
    """
    asked = session.last_asked_field
    if asked is None or asked not in REQUIRED_FIELDS:
        return extraction, None

    field = cast(BriefFieldName, asked)
    kept: list[FieldUpdate] = []
    rejected_asked = False
    for update in extraction.updates:
        if update.field != field:
            kept.append(update)
            continue
        if _is_acceptable_answer(field, update):
            kept.append(update)
        else:
            rejected_asked = True

    if not rejected_asked:
        return extraction, None

    return extraction.model_copy(update={"updates": kept}), field

File: backend/src/orchestration/answer_guard.ts
import { meaningfulFeatures } from "../domain/gate.ts";
import {
  type BriefFieldName,
  type ExtractionResult,
  type FieldUpdate,
  type Session,
  REQUIRED_FIELDS,
} from "../domain/models.ts";

const PRICE_CONFIRMED_RE = /\d|(?:pln|usd|eur|gbp|zł)\b|[$€£]/i;
const PRICE_VAGUE_RE =
  /\b(cheap|cheaper|cheapest|expensive|affordable|budget|pricey|inexpensive|costly|reasonable|low[\s-]?cost|high[\s-]?end|around|about|approximately|roughly|starting|from)\b/i;

function priceText(update: FieldUpdate): string {
  if (update.status === "confirmed" && update.value !== null) {
    if (Array.isArray(update.value)) {
      return update.value.map((item) => String(item)).join(" ");
    }
    return String(update.value);
  }
  return update.raw_text ?? "";
}

function isAcceptablePrice(update: FieldUpdate): boolean {
  const text = priceText(update).trim();
  if (!text) {
    return false;
  }
  if (update.status === "confirmed") {
    return PRICE_CONFIRMED_RE.test(text);
  }
  return PRICE_VAGUE_RE.test(text);
}

function isAcceptableAnswer(field: BriefFieldName, update: FieldUpdate): boolean {
  if (field === "price") {
    return isAcceptablePrice(update);
  }
  if (update.status !== "confirmed") {
    return false;
  }
  if (update.value === null) {
    return false;
  }
  if (field === "key_features") {
    return meaningfulFeatures(update.value).length > 0;
  }
  if (typeof update.value === "string") {
    return Boolean(update.value.trim());
  }
  if (Array.isArray(update.value)) {
    return meaningfulFeatures(update.value).length > 0;
  }
  return true;
}

export function filterInvalidAskedFieldUpdates(
  session: Session,
  extraction: ExtractionResult,
): { extraction: ExtractionResult; rejectedField: BriefFieldName | null } {
  const asked = session.last_asked_field;
  if (asked === null || !REQUIRED_FIELDS.includes(asked as BriefFieldName)) {
    return { extraction, rejectedField: null };
  }

  const field = asked as BriefFieldName;
  const kept: FieldUpdate[] = [];
  let rejectedAsked = false;
  for (const update of extraction.updates) {
    if (update.field !== field) {
      kept.push(update);
      continue;
    }
    if (isAcceptableAnswer(field, update)) {
      kept.push(update);
    } else {
      rejectedAsked = true;
    }
  }

  if (!rejectedAsked) {
    return { extraction, rejectedField: null };
  }

  return {
    extraction: { ...extraction, updates: kept },
    rejectedField: field,
  };
}
